edb/server/http/src/python.rs

```diff
@@ -0,0 +1,434 @@
+        info!("Http dropped");
+    }
+}
+
+fn internal_error(message: &str) -> PyErr {
+    error!("{message}");
+    InternalError::new_err(())
+}
+
+async fn request(
+    client: reqwest::Client,
+    url: String,
+    method: String,
+    body: Vec<u8>,
+    headers: Vec<(String, String)>,
+) -> Result<(reqwest::StatusCode, Vec<u8>, HashMap<String, String>), String> {
+    let method =
+        Method::from_bytes(method.as_bytes()).map_err(|e| format!("Invalid HTTP method: {}", e))?;
+
+    let mut req = client.request(method, url);
+
+    for (key, value) in headers {
+        req = req.header(key, value);
+    }
+
+    if !body.is_empty() {
+        req = req.body(body);
+    }
+
+    let resp = req
+        .send()
+        .await
+        .map_err(|e| format!("Request failed: {}", e))?;
+
+    let status = resp.status();
+
+    let headers = resp
+        .headers()
+        .iter()
+        .map(|(k, v)| (k.to_string(), v.to_str().unwrap_or("").to_string()))
+        .collect();
+
+    let body = resp
+        .bytes()
+        .await
+        .map_err(|e| format!("Failed to read response body: {}", e))?
+        .to_vec();
+
+    Ok((status, body, headers))
+}
+
+struct PermitCount {
+    active: usize,
+    capacity: usize,
+}
+
+/// By default, the [`Semaphore`] does not allow for releasing permits that are currently
+/// outstanding. This allows us to keep a count of what the limit _should_ be, which we'll
+/// target if we've taken too many permits out by forgetting outstanding permits.
+struct PermitManager {
+    counts: Mutex<PermitCount>,
+    semaphore: Semaphore,
+}
+
+impl PermitManager {
+    fn new(capacity: usize) -> Self {
+        Self {
+            counts: Mutex::new(PermitCount {
+                active: 0,
+                capacity,
+            }),
+            semaphore: Semaphore::new(capacity),
+        }
+    }
+
+    async fn acquire(&self) -> Result<SemaphorePermit<'_>, AcquireError> {
+        let permit = self.semaphore.acquire().await?;
+        let mut counts = self.counts.lock().unwrap();
+        counts.active += 1;
+        Ok(permit)
+    }
+
+    fn release(&self, permit: SemaphorePermit<'_>) {
+        let mut counts = self.counts.lock().unwrap();
+        if counts.active > counts.capacity {
+            // We have too many permits, so forget this one instead of releasing
+            permit.forget();
+        } else {
+            // Normal release
+            drop(permit);
+        }
+        counts.active -= 1;
+    }
+
+    fn update_limit(&self, new_limit: usize) {
+        let mut counts = self.counts.lock().unwrap();
+        let old_capacity = counts.capacity;
+        counts.capacity = new_limit;
+
+        if new_limit > old_capacity {
+            let added = new_limit - old_capacity;
+            self.semaphore.add_permits(added);
+        } else if new_limit < old_capacity {
+            let removed = old_capacity - new_limit;
+            self.semaphore.forget_permits(removed);
```

Comment:
Add a comment that this might forget too few permits, but we make up for it in release?

Reply:
Oh, I think there might actually be a race here.
Imagine that `capacity` is 10 and `active` is 6, so there are 4 permits remaining.
We call `update_limit(5)`, and try to forget `5` permits, but only `4` actually get forgotten.
Now capacity is 5, active is 6, and there are 0 permits remaining.
If `update_limit(10)` gets called now, we'll call `add_permits(5)` and update `capacity` to 10.
Then, when the 6 active threads release, they will all return their permits (since they are no longer over the limit), and we will wind up with 11 permits in the semaphore.

Reply:
Ack. I think you are correct. Let me add a test for this case.

Reply:
Added a bunch more validation, and some extensive tests around acquire/release that run only in the `debug_assertions` cfg.